dmenu-rs: refactor to add variant with enabled plugins

benjaminedwardwebb opened the pull request:
## Description of changes

Add the `dmenu-rs-enable-plugins` package.

The `dmenu-rs-enable-plugins` package is identical the `dmenu-rs`, except
that it is built with all plugins available in the `dmenu-rs` respository
(whereas the `dmenu-rs` package is built with none). The same underlying
derivation is use, except a boolean flag is passed to indicate whether
or not to enable plugins.

This commit also moves the `dmenu-rs` package into `pkgs/by-name`.

See https://github.com/NixOS/nixpkgs/issues/333733
and https://github.com/Shizcow/dmenu-rs

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

GaetanLepage commented on pkgs/by-name/dm/dmenu-rs-full/package.nix:
```diff
@@ -0,0 +1,92 @@
+# plugins from outside of the dmenu-rs repository, you'll have to build it
+# from the source.
+# See: https://github.com/Shizcow/dmenu-rs#plugins
+stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation rec {
```
Any reason to not use `rustPlatform.buildRustPackage` ?

GaetanLepage commented on pkgs/by-name/dm/dmenu-rs-full/package.nix:
```diff
@@ -0,0 +1,92 @@
+  src = fetchFromGitHub {
+    owner = "Shizcow";
+    repo = "dmenu-rs";
+    rev = version;
```
```suggestion
    rev = "refs/tags/${version}";
```

Avoids branch-tag mismatch.

GaetanLepage commented on pkgs/by-name/dm/dmenu-rs-full/package.nix:
```diff
@@ -0,0 +1,92 @@
+  # deterministically. See the original PR for some discussion on this.
+  doCheck = false;
+
+  meta = with lib; {
+    description = "A pixel perfect port of dmenu, rewritten in Rust with extensive plugin support (plugins enabled)";
+    homepage = "https://github.com/Shizcow/dmenu-rs";
+    license = with licenses; [ gpl3Only ];
+    maintainers = with maintainers; [ benjaminedwardwebb ];
+    platforms = platforms.linux;
+    broken = (stdenv.isLinux && stdenv.isAarch64);
+  };
```
```suggestion
  meta = {
    description = "A pixel perfect port of dmenu, rewritten in Rust with extensive plugin support (plugins enabled)";
    homepage = "https://github.com/Shizcow/dmenu-rs";
    license = with lib.licenses; [ gpl3Only ];
    maintainers = with lib.maintainers; [ benjaminedwardwebb ];
    platforms = lib.platforms.linux;
    broken = (stdenv.isLinux && stdenv.isAarch64);
  };
```

GaetanLepage reviewed (changes requested):
Minor comments for now

GaetanLepage commented:
> Yeah, I think refactoring the packages so that they rely on the same underlying, parameterized nix build is a great idea! Let me do that -- and thank you for chiming in. I kind of just didn't know what the conventions were around that, so kept things simple and "copy-paste" until someone suggested a better way.
> 
> However, I think we'll need two `Cargo.lock` files regardless. The plugins are defined at build-time and the selection of plugins are used in the upstream `make` build to dynamically generate the proper `Cargo.lock` file (hence why I opted to check it in at all originally). There's probably a better, clever way to work with dynamic `Cargo.lock` files with nixpkgs' rust tooling, but I haven't been able to figure it out.

Looks good ! Feel free to ping me whenever this will be ready.

GaetanLepage commented on pkgs/by-name/dm/dmenu-rs/package.nix:
```diff
@@ -0,0 +1,118 @@
+    '' else ""
+  );
+
+  preBuild = if enablePlugins
+    # Include all plugins in the dmenu-rs repository under src/plugins.
+    # See https://github.com/Shizcow/dmenu-rs/tree/master/src/plugins
+    then ''
+      sed -i -E "s/PLUGINS =/PLUGINS = $(find src/plugins/ -mindepth 1 -maxdepth 1 -type d -printf "%f\n" | tr "\n" " ")/" config.mk
+    ''
+    else "";
```
```suggestion
  # Include all plugins in the dmenu-rs repository under src/plugins.
  # See https://github.com/Shizcow/dmenu-rs/tree/master/src/plugins
  preBuild = lib.optionalString enablePlugins ''
      sed -i -E "s/PLUGINS =/PLUGINS = $(find src/plugins/ -mindepth 1 -maxdepth 1 -type d -printf "%f\n" | tr "\n" " ")/" config.mk
  '';
```

GaetanLepage commented on pkgs/by-name/dm/dmenu-rs/package.nix:
```diff
@@ -0,0 +1,118 @@
+  meta = {
+    description =
+      "A pixel perfect port of dmenu, rewritten in Rust with extensive plugin support"
+      + (
+        if enablePlugins
+        then ", with all upstream plugins enabled"
+        else ""
+      );
```
```suggestion
      + lib.optionalString enablePlugins ", with all upstream plugins enabled";
```

GaetanLepage commented on pkgs/by-name/dm/dmenu-rs/package.nix:
```diff
@@ -0,0 +1,114 @@
+      libXft
+      libXinerama
+    ]
+    ++ (
+      if enablePlugins then
+        [
+          aspell
+          xclip
+          xdg-utils
+        ]
+      else
+        [ ]
+    );
```
```suggestion
    ++ lib.optionals enablePlugins [
      aspell
      xclip
      xdg-utils
    ];
```

GaetanLepage reviewed (approved):
LGTM

SuperSandro2000 commented on pkgs/by-name/dm/dmenu-rs/package.nix:
```diff
@@ -0,0 +1,109 @@
+
+  meta = {
+    description =
+      "A pixel perfect port of dmenu, rewritten in Rust with extensive plugin support"
```
```suggestion
      "Pixel perfect port of dmenu, rewritten in Rust with extensive plugin support"
```

GaetanLepage replied:
Sure, but [we have some guidelines that often need you to adapt the upstream description](https://nixos.org/manual/nixpkgs/unstable/#var-meta-description).

GaetanLepage replied:
This manual is a great resource for contributing to nixpkgs.
Obviously, reading it from top to bottom is not super feasible, but a few `Ctrl-f` from time to time really help.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28672,7 +28672,7 @@ with pkgs;
   dmenu-wayland = callPackage ../applications/misc/dmenu/wayland.nix { };
 
-  dmenu-rs = callPackage ../applications/misc/dmenu-rs { };
+  dmenu-rs-enable-plugins = callPackage ../by-name/dm/dmenu-rs/package.nix { enablePlugins = true; };
```
```suggestion
  dmenu-rs-enable-plugins = dmenu-rs.override { enablePlugins = true; };
```

GaetanLepage commented on pkgs/by-name/dm/dmenu-rs/package.nix:
```diff
@@ -0,0 +1,109 @@
+    license = with lib.licenses; [ gpl3Only ];
+    maintainers = with lib.maintainers; [ benjaminedwardwebb ];
+    platforms = lib.platforms.linux;
+    broken = (stdenv.isLinux && stdenv.isAarch64);
```
```suggestion
    broken = stdenv.hostPlatform.isLinux && stdenv.hostPlatform.isAarch64;
```

GaetanLepage commented on pkgs/by-name/dm/dmenu-rs/package.nix:
```diff
@@ -0,0 +1,109 @@
+{
+  aspell,
+  cargo,
+  expat,
+  fetchFromGitHub,
+  fontconfig,
+  lib,
+  libXft,
+  libXinerama,
+  m4,
+  pkg-config,
+  python3,
+  rustPlatform,
+  stdenv,
+  xclip,
+  xdg-utils,
+  enablePlugins ? false,
+}:
```
```suggestion
{
  lib,
  stdenv,
  fetchFromGitHub,

  # nativeBuildInputs
  cargo,
  m4,
  pkg-config,
  python3,
  rustPlatform,

  # buildInputs
  expat,
  fontconfig,
  libXft,
  libXinerama,
  aspell,
  xclip,
  xdg-utils,

  enablePlugins ? false,
}:
```
Arguments should be sorted in the appearing order. Also, `lib` and `stdenv` should go at the top.

GaetanLepage replied:
I personally prefer when the sections are separated, but you are free to ignore this suggestion. This is more subjective.

GaetanLepage replied:
> Do you prefer the sections to have a comment (e.g., `nativeBuildInputs`) as well?

I tend to do so when I package things. Again, this is quite subjective.

GaetanLepage commented on pkgs/by-name/dm/dmenu-rs/package.nix:
```diff
@@ -0,0 +1,114 @@
+    license = with lib.licenses; [ gpl3Only ];
+    maintainers = with lib.maintainers; [ benjaminedwardwebb ];
+    platforms = lib.platforms.linux;
+    broken = (stdenv.hostPlatform.isLinux && stdenv.hostPlatform.isAarch64);
```
No need for the parentheses here.

GaetanLepage reviewed (approved):
LGTM, thank you for taking the time to address all of my feedback.
@SuperSandro2000 I let you have a last look and press the merge button :)

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 333837`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>dmenu-rs</li>
    <li>dmenu-rs-enable-plugins</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>dmenu-rs</li>
    <li>dmenu-rs-enable-plugins</li>
  </ul>
</details>

Aleksanaa reviewed (commented):
Please separate `dmenu-rs: move to pkgs/by-name`, or people will have problems trace original files. Others seems fine

benjaminedwardwebb committed:
dmenu-rs: move to pkgs/by-name

benjaminedwardwebb committed:
dmenu-rs: refactor to add variant with enabled plugins

Add the dmenu-rs-enable-plugins package.

The dmenu-rs-enable-plugins package is identical the dmenu-rs, except
that it is built with all plugins available in the dmenu-rs respository
(whereas the dmenu-rs package is built with none). The same underlying
derivation is use, except a boolean flag is passed to indicate whether
or not to enable plugins.

See https://github.com/NixOS/nixpkgs/issues/333733
and https://github.com/Shizcow/dmenu-rs

GaetanLepage reviewed (approved):
LGTM
